demos/twa-webview-fallback/src/main/java/com/google/browser/examples/twawebviewfallback/WebViewFallbackActivity.java

```diff
@@ -0,0 +1,106 @@
+import androidx.browser.customtabs.CustomTabsIntent;
+
+import android.net.Uri;
+import android.os.Bundle;
+import android.view.View;
+import android.view.ViewGroup;
+import android.webkit.RenderProcessGoneDetail;
+import android.webkit.WebResourceRequest;
+import android.webkit.WebSettings;
+import android.webkit.WebView;
+import android.webkit.WebViewClient;
+import android.widget.FrameLayout;
+import android.widget.Toast;
+
+public class WebViewFallbackActivity extends AppCompatActivity {
+    public static final String KEY_LAUNCH_URI =
+            "com.google.browser.examples.twawebviewfallback.WebViewFallbackActivity.KEY_LAUNCH_URL";
+
+    private Uri launchUrl;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        this.launchUrl = this.getIntent().getParcelableExtra(KEY_LAUNCH_URI);
+
+        FrameLayout webViewContainer = new FrameLayout(this);
+        WebView webView = new WebView(this);
+        webView.setWebViewClient(createWebViewClient());
+        webViewContainer.addView(webView,
+                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
+
+        WebSettings webSettings = webView.getSettings();
+        webSettings.setJavaScriptEnabled(true);
+
+        FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(
+                FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT);
+
+        setContentView(webViewContainer,layoutParams);
+        webView.loadUrl(launchUrl.toString());
+    }
+
+    private WebViewClient createWebViewClient() {
+        return new WebViewClient() {
+            @Override
+            public boolean onRenderProcessGone(
+                    WebView view, RenderProcessGoneDetail detail) {
+                ViewGroup vg = (ViewGroup)view.getParent();
+
+                // Remove crashed WebView from the hierarchy
+                // and ensure it is destroyed.
+                vg.removeView(view);
+                view.destroy();
+
+                // Create a new instance, and ensure it also
+                // handles crashes - in this case, re-using
+                // the current WebViewClient
+                WebView webView = new WebView(view.getContext());
+                webView.setWebViewClient(this);
+                vg.addView(webView);
+
+                // With the crash recovered, decide what to do next.
+                // We are sending a toast and loading the origin
+                // URL, in this example.
+                Toast.makeText(view.getContext(), "Recovering from crash",
+                        Toast.LENGTH_LONG).show();
+                webView.loadUrl(launchUrl.toString());
+                return true;
+            }
+
+            @Override
+            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
+                Uri launchUrl = WebViewFallbackActivity.this.launchUrl;
```

Comment:
This will break in the event that someone creates a TWA that launches to a site off the verified origin. I don't know why they'd do that, so maybe we could just say they're holding it wrong at that point.

Reply:
Actually, we are not validating origins at all in the fallback. It seems worth doing it in another PR.